## Ted's Analysis

Please fill out:
* Student name: 
* Student pace: self paced / part time / full time
* Scheduled project review date/time: 
* Instructor name: 
* Blog post URL:


In [1]:
# Your code here - remember to use markdown cells for comments as well!



In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from datetime import datetime
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.impute import MissingIndicator
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OrdinalEncoder
from sklearn.preprocessing import OneHotEncoder
from sklearn.model_selection import cross_val_score


In [3]:
df = pd.read_csv('data/kc_house_data.csv')
df

id        date     price  bedrooms  bathrooms  sqft_living  \
0      7129300520  10/13/2014  221900.0         3       1.00         1180   
1      6414100192   12/9/2014  538000.0         3       2.25         2570   
2      5631500400   2/25/2015  180000.0         2       1.00          770   
3      2487200875   12/9/2014  604000.0         4       3.00         1960   
4      1954400510   2/18/2015  510000.0         3       2.00         1680   
...           ...         ...       ...       ...        ...          ...   
21592   263000018   5/21/2014  360000.0         3       2.50         1530   
21593  6600060120   2/23/2015  400000.0         4       2.50         2310   
21594  1523300141   6/23/2014  402101.0         2       0.75         1020   
21595   291310100   1/16/2015  400000.0         3       2.50         1600   
21596  1523300157  10/15/2014  325000.0         2       0.75         1020   

       sqft_lot  floors waterfront  view  ...          grade sqft_above  \
0          5650     1.0        NaN  NONE  ...      7 Average       1180   
1          7242     2.0         NO  NONE  ...      7 Average       2170   
2         10000     1.0         NO  NONE  ...  6 Low Average        770   
3          5000     1.0         NO  NONE  ...      7 Average       1050   
4          8080     1.0         NO  NONE  ...         8 Good       1680   
...         ...     ...        ...   ...  ...            ...        ...   
21592      1131     3.0         NO  NONE  ...         8 Good       1530   
21593      5813     2.0         NO  NONE  ...         8 Good       2310   
21594      1350     2.0         NO  NONE  ...      7 Average       1020   
21595      2388     2.0        NaN  NONE  ...         8 Good       1600   
21596      1076     2.0         NO  NONE  ...      7 Average       1020   

       sqft_basement yr_built  yr_renovated  zipcode      lat     long  \
0                0.0     1955           0.0    98178  47.5112 -122.257   
1              400.0     1951        1991.0    98125  47.7210 -122.319   
2                0.0     1933           NaN    98028  47.7379 -122.233   
3              910.0     1965           0.0    98136  47.5208 -122.393   
4                0.0     1987           0.0    98074  47.6168 -122.045   
...              ...      ...           ...      ...      ...      ...   
21592            0.0     2009           0.0    98103  47.6993 -122.346   
21593            0.0     2014           0.0    98146  47.5107 -122.362   
21594            0.0     2009           0.0    98144  47.5944 -122.299   
21595            0.0     2004           0.0    98027  47.5345 -122.069   
21596            0.0     2008           0.0    98144  47.5941 -122.299   

       sqft_living15  sqft_lot15  
0               1340        5650  
1               1690        7639  
2               2720        8062  
3               1360        5000  
4               1800        7503  
...              ...         ...  
21592           1530        1509  
21593           1830        7200  
21594           1020        2007  
21595           1410        1287  
21596           1020        1357  

[21597 rows x 21 columns]

In [4]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.159700e+04  2.159700e+04  21597.000000  21597.000000  21597.000000   
mean   4.580474e+09  5.402966e+05      3.373200      2.115826   2080.321850   
std    2.876736e+09  3.673681e+05      0.926299      0.768984    918.106125   
min    1.000102e+06  7.800000e+04      1.000000      0.500000    370.000000   
25%    2.123049e+09  3.220000e+05      3.000000      1.750000   1430.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    sqft_above      yr_built  yr_renovated  \
count  2.159700e+04  21597.000000  21597.000000  21597.000000  17755.000000   
mean   1.509941e+04      1.494096   1788.596842   1970.999676     83.636778   
std    4.141264e+04      0.539683    827.759761     29.375234    399.946414   
min    5.200000e+02      1.000000    370.000000   1900.000000      0.000000   
25%    5.040000e+03      1.000000   1190.000000   1951.000000      0.000000   
50%    7.618000e+03      1.500000   1560.000000   1975.000000      0.000000   
75%    1.068500e+04      2.000000   2210.000000   1997.000000      0.000000   
max    1.651359e+06      3.500000   9410.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21597.000000  21597.000000  21597.000000   21597.000000   21597.000000  
mean   98077.951845     47.560093   -122.213982    1986.620318   12758.283512  
std       53.513072      0.138552      0.140724     685.230472   27274.441950  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471100   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.231000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [5]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

# The prediction target is price

In [7]:
y = df['price']
X = df.drop('price',axis=1)

In [8]:
##  sklearn model_selection train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=69)

In [9]:
print(f"X_train is a DataFram with {X_train.shape[0]} rows and {X_train.shape[1]} columns")
print(f"y_train is a Series with {y_train.shape[0]} values")

X_train is a DataFram with 16197 rows and 20 columns
y_train is a Series with 16197 values


In [10]:
print(f"X_test is a DataFram with {X_test.shape[0]} rows and {X_test.shape[1]} columns")
print(f"y_test is a Series with {y_test.shape[0]} values")

X_test is a DataFram with 5400 rows and 20 columns
y_test is a Series with 5400 values


In [11]:
##  sklearn linear_model LinearRegression:  Should fail at this point and it does, so it is
##    commented out.  This includes ALL, non-converted data, so it needs lots of cleaning

# model = LinearRegression()
# model.fit(X_train, y_train)

In [12]:
X_train

id        date  bedrooms  bathrooms  sqft_living  sqft_lot  \
17085  2919201365   6/16/2014         4       2.75         2610      4160   
13333  1822350180  12/11/2014         3       2.25         1330      8004   
12562   302000065   1/29/2015         3       1.00          970     14850   
11348  1795500060  10/21/2014         3       1.00         1040      8645   
10228   179000505  10/24/2014         3       1.75         1800      9000   
...           ...         ...       ...        ...          ...       ...   
4374    321059059   5/19/2014         3       1.00         1290    189486   
11825  4139910160    4/1/2015         5       3.25         4320     32840   
14740  2423039122   7/18/2014         4       1.00         1900      9000   
9818   2550820060   4/28/2015         3       1.75         1630     10001   
4041   9834200165    4/6/2015         4       1.50         1790      4080   

       floors waterfront  view  condition      grade  sqft_above  \
17085     3.0         NO  NONE  Very Good     8 Good        1910   
13333     2.0         NO  NONE    Average  7 Average        1330   
12562     1.0         NO  NONE    Average  7 Average         970   
11348     1.0         NO  NONE       Good  7 Average        1040   
10228     1.0         NO  NONE    Average  7 Average        1200   
...       ...        ...   ...        ...        ...         ...   
4374      1.0         NO  NONE       Good  7 Average        1290   
11825     2.0         NO  NONE    Average  12 Luxury        4320   
14740     1.0        NaN  NONE       Good  7 Average        1290   
9818      1.0         NO  NONE       Good  7 Average        1100   
4041      1.5         NO  NONE  Very Good  7 Average        1790   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
17085         700.0      1910           0.0    98103  47.6901 -122.357   
13333           0.0      1985           0.0    98034  47.7098 -122.217   
12562             ?      1968           0.0    98001  47.3251 -122.268   
11348           0.0      1962           NaN    98042  47.3631 -122.116   
10228         600.0      1961           NaN    98178  47.4932 -122.278   
...             ...       ...           ...      ...      ...      ...   
4374            0.0      1960           0.0    98092  47.3356 -122.157   
11825           0.0      1990           NaN    98006  47.5461 -122.122   
14740         610.0      1948           NaN    98166  47.4629 -122.361   
9818          530.0      1977           0.0    98042  47.3605 -122.120   
4041            0.0      1928           0.0    98144  47.5749 -122.291   

       sqft_living15  sqft_lot15  
17085           1470        4140  
13333           1300        7971  
12562           1410       14850  
11348           1290        8645  
10228            980        6000  
...              ...         ...  
4374            2370       98881  
11825           4410       33210  
14740           1950       10800  
9818            1630       10001  
4041            1710        4080  

[16197 rows x 20 columns]

###  Converting the sell date ("date") to a sortable integer

In [13]:
# ##  Change date format, but keep, in case we want to use it later

date_num = []
for date in X_train['date']:
    split = [char for char in date]
    year = split[-4] + split[-3] + split[-2] + split[-1]
    if isinstance(split[1],int):
        month = split[0] + split[1]
    else:
        month = split[0]
    if len(month) == 1:
        date_num.append(int(year + '0' + month))
    else:
        date_num.append(int(year + month))
X_train['date_num'] = date_num

<ipython-input-13-74e9c2404cf9>:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  X_train['date_num'] = date_num


#### Dropping the sell date so a regression can still be performed

In [14]:
###  Date and other columns to drop(lat and long) - not useful for this analysis
lat_long = ['lat','long','date']
X_train = X_train.drop(lat_long, axis=1)
X_train

id  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
17085  2919201365         4       2.75         2610      4160     3.0   
13333  1822350180         3       2.25         1330      8004     2.0   
12562   302000065         3       1.00          970     14850     1.0   
11348  1795500060         3       1.00         1040      8645     1.0   
10228   179000505         3       1.75         1800      9000     1.0   
...           ...       ...        ...          ...       ...     ...   
4374    321059059         3       1.00         1290    189486     1.0   
11825  4139910160         5       3.25         4320     32840     2.0   
14740  2423039122         4       1.00         1900      9000     1.0   
9818   2550820060         3       1.75         1630     10001     1.0   
4041   9834200165         4       1.50         1790      4080     1.5   

      waterfront  view  condition      grade  sqft_above sqft_basement  \
17085         NO  NONE  Very Good     8 Good        1910         700.0   
13333         NO  NONE    Average  7 Average        1330           0.0   
12562         NO  NONE    Average  7 Average         970             ?   
11348         NO  NONE       Good  7 Average        1040           0.0   
10228         NO  NONE    Average  7 Average        1200         600.0   
...          ...   ...        ...        ...         ...           ...   
4374          NO  NONE       Good  7 Average        1290           0.0   
11825         NO  NONE    Average  12 Luxury        4320           0.0   
14740        NaN  NONE       Good  7 Average        1290         610.0   
9818          NO  NONE       Good  7 Average        1100         530.0   
4041          NO  NONE  Very Good  7 Average        1790           0.0   

       yr_built  yr_renovated  zipcode  sqft_living15  sqft_lot15  date_num  
17085      1910           0.0    98103           1470        4140    201406  
13333      1985           0.0    98034           1300        7971    201401  
12562      1968           0.0    98001           1410       14850    201501  
11348      1962           NaN    98042           1290        8645    201401  
10228      1961           NaN    98178            980        6000    201401  
...         ...           ...      ...            ...         ...       ...  
4374       1960           0.0    98092           2370       98881    201405  
11825      1990           NaN    98006           4410       33210    201504  
14740      1948           NaN    98166           1950       10800    201407  
9818       1977           0.0    98042           1630       10001    201504  
4041       1928           0.0    98144           1710        4080    201504  

[16197 rows x 18 columns]

In [15]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16197 entries, 17085 to 4041
Data columns (total 18 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             16197 non-null  int64  
 1   bedrooms       16197 non-null  int64  
 2   bathrooms      16197 non-null  float64
 3   sqft_living    16197 non-null  int64  
 4   sqft_lot       16197 non-null  int64  
 5   floors         16197 non-null  float64
 6   waterfront     14375 non-null  object 
 7   view           16148 non-null  object 
 8   condition      16197 non-null  object 
 9   grade          16197 non-null  object 
 10  sqft_above     16197 non-null  int64  
 11  sqft_basement  16197 non-null  object 
 12  yr_built       16197 non-null  int64  
 13  yr_renovated   13292 non-null  float64
 14  zipcode        16197 non-null  int64  
 15  sqft_living15  16197 non-null  int64  
 16  sqft_lot15     16197 non-null  int64  
 17  date_num       16197 non-null  int64  
dtypes: 

In [16]:
X_train['waterfront'].value_counts()

NO     14270
YES      105
Name: waterfront, dtype: int64

In [17]:
X_train.isna().sum()

id                  0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront       1822
view               49
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     2905
zipcode             0
sqft_living15       0
sqft_lot15          0
date_num            0
dtype: int64

Since there are 1822 listing with "waterfront" missing, but less than 1% of listed homes have a waterfront access, these seem like infrequent enough cases that we will fill in the waterfront NaNs with "NO" instead

In [18]:
X_train['waterfront'] = X_train['waterfront'].fillna("NO")

In [19]:
X_train.isna().sum()

id                  0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront          0
view               49
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     2905
zipcode             0
sqft_living15       0
sqft_lot15          0
date_num            0
dtype: int64

Doing the same thing with the "view column now:

In [20]:
X_train['view'].value_counts()

NONE         14534
AVERAGE        745
GOOD           384
FAIR           252
EXCELLENT      233
Name: view, dtype: int64

We will go with the mode of this to replace the 49 missing values, as this is the likely case.

In [21]:
X_train['view'] = X_train['view'].fillna("NONE")

In [22]:
X_train.isna().sum()

id                  0
bedrooms            0
bathrooms           0
sqft_living         0
sqft_lot            0
floors              0
waterfront          0
view                0
condition           0
grade               0
sqft_above          0
sqft_basement       0
yr_built            0
yr_renovated     2905
zipcode             0
sqft_living15       0
sqft_lot15          0
date_num            0
dtype: int64

In [26]:
X_train['yr_renovated'].value_counts()

0.0       12732
2014.0       56
2003.0       23
2000.0       22
2013.0       20
          ...  
1951.0        1
1946.0        1
1948.0        1
1954.0        1
1944.0        1
Name: yr_renovated, Length: 66, dtype: int64

Very few homes from this list have been renovated.  The 0 values may have never been renovated, but we can't know that for sure.  Since there are so many 0s and so many missing entries, the column itself needs to be dropped.  If we want to add it back later, we'll have to address these issues.

In [29]:
X_train = X_train.drop('yr_renovated', axis=1)
X_train

id  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
17085  2919201365         4       2.75         2610      4160     3.0   
13333  1822350180         3       2.25         1330      8004     2.0   
12562   302000065         3       1.00          970     14850     1.0   
11348  1795500060         3       1.00         1040      8645     1.0   
10228   179000505         3       1.75         1800      9000     1.0   
...           ...       ...        ...          ...       ...     ...   
4374    321059059         3       1.00         1290    189486     1.0   
11825  4139910160         5       3.25         4320     32840     2.0   
14740  2423039122         4       1.00         1900      9000     1.0   
9818   2550820060         3       1.75         1630     10001     1.0   
4041   9834200165         4       1.50         1790      4080     1.5   

      waterfront  view  condition      grade  sqft_above sqft_basement  \
17085         NO  NONE  Very Good     8 Good        1910         700.0   
13333         NO  NONE    Average  7 Average        1330           0.0   
12562         NO  NONE    Average  7 Average         970             ?   
11348         NO  NONE       Good  7 Average        1040           0.0   
10228         NO  NONE    Average  7 Average        1200         600.0   
...          ...   ...        ...        ...         ...           ...   
4374          NO  NONE       Good  7 Average        1290           0.0   
11825         NO  NONE    Average  12 Luxury        4320           0.0   
14740         NO  NONE       Good  7 Average        1290         610.0   
9818          NO  NONE       Good  7 Average        1100         530.0   
4041          NO  NONE  Very Good  7 Average        1790           0.0   

       yr_built  zipcode  sqft_living15  sqft_lot15  date_num  
17085      1910    98103           1470        4140    201406  
13333      1985    98034           1300        7971    201401  
12562      1968    98001           1410       14850    201501  
11348      1962    98042           1290        8645    201401  
10228      1961    98178            980        6000    201401  
...         ...      ...            ...         ...       ...  
4374       1960    98092           2370       98881    201405  
11825      1990    98006           4410       33210    201504  
14740      1948    98166           1950       10800    201407  
9818       1977    98042           1630       10001    201504  
4041       1928    98144           1710        4080    201504  

[16197 rows x 17 columns]

In [30]:
X_train.isna().sum()

id               0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
view             0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
zipcode          0
sqft_living15    0
sqft_lot15       0
date_num         0
dtype: int64

In [32]:
X_train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16197 entries, 17085 to 4041
Data columns (total 17 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             16197 non-null  int64  
 1   bedrooms       16197 non-null  int64  
 2   bathrooms      16197 non-null  float64
 3   sqft_living    16197 non-null  int64  
 4   sqft_lot       16197 non-null  int64  
 5   floors         16197 non-null  float64
 6   waterfront     16197 non-null  object 
 7   view           16197 non-null  object 
 8   condition      16197 non-null  object 
 9   grade          16197 non-null  object 
 10  sqft_above     16197 non-null  int64  
 11  sqft_basement  16197 non-null  object 
 12  yr_built       16197 non-null  int64  
 13  zipcode        16197 non-null  int64  
 14  sqft_living15  16197 non-null  int64  
 15  sqft_lot15     16197 non-null  int64  
 16  date_num       16197 non-null  int64  
dtypes: float64(2), int64(10), object(5)
memory usag

In [33]:
X_train.describe()

id      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  1.619700e+04  16197.000000  16197.000000  16197.000000  1.619700e+04   
mean   4.582690e+09      3.374143      2.119034   2088.038834  1.524385e+04   
std    2.876658e+09      0.900683      0.770777    925.153633  4.197362e+04   
min    1.000102e+06      1.000000      0.500000    370.000000  5.200000e+02   
25%    2.122049e+09      3.000000      1.750000   1430.000000  5.070000e+03   
50%    3.905000e+09      3.000000      2.250000   1920.000000  7.620000e+03   
75%    7.312400e+09      4.000000      2.500000   2560.000000  1.072000e+04   
max    9.900000e+09     11.000000      8.000000  13540.000000  1.651359e+06   

             floors    sqft_above      yr_built       zipcode  sqft_living15  \
count  16197.000000  16197.000000  16197.000000  16197.000000   16197.000000   
mean       1.492993   1795.619004   1970.954251  98078.094462    1990.863493   
std        0.538724    835.426559     29.404289     53.453526     691.165364   
min        1.000000    370.000000   1900.000000  98001.000000     399.000000   
25%        1.000000   1200.000000   1951.000000  98033.000000    1490.000000   
50%        1.500000   1570.000000   1975.000000  98065.000000    1840.000000   
75%        2.000000   2230.000000   1997.000000  98118.000000    2370.000000   
max        3.500000   9410.000000   2015.000000  98199.000000    6210.000000   

          sqft_lot15       date_num  
count   16197.000000   16197.000000  
mean    12952.254368  201436.421992  
std     28093.879666      45.795963  
min       651.000000  201401.000000  
25%      5100.000000  201405.000000  
50%      7623.000000  201407.000000  
75%     10083.000000  201502.000000  
max    871200.000000  201505.000000